nvme_gw_server.py

```diff
@@ -528,74 +532,96 @@ def nvmf_subsystem_remove_host(self, request, context=None):
     def nvmf_subsystem_add_listener(self, request, context=None):
         """Adds a listener at the given TCP/IP address for the given subsystem."""
         self.logger.info({
-            f"Adding {request.trtype} listener at {request.traddr} : {request.trsvcid} for {request.nqn}"
+            f"Adding {request.gateway} {request.trtype} listener at {request.traddr} : {request.trsvcid} for {request.nqn}"
         })
 
-        try:
-            if request.trtype.lower() not in self.transports:
-                raise Exception(f"{request.trtype} transport is not enabled")
+        if not request.gateway or request.gateway == self.gateway_name:
+            try:
+                traddr = request.traddr
+                if not request.gateway:
+                    if traddr:
+                        raise Exception(f"{request.traddr} address reqires gateway")
+                    traddr = self.nvme_config.get("config", "gateway_addr")
 
-            return_string = self.spdk_rpc.nvmf.nvmf_subsystem_add_listener(
-                self.spdk_rpc_client,
-                nqn=request.nqn,
-                trtype=request.trtype,
-                traddr=request.traddr,
-                trsvcid=request.trsvcid,
-                adrfam=request.adrfam,
-            )
-            self.logger.info(f"Status of add listener: {return_string}")
-        except Exception as ex:
-            self.logger.error(f"Add Listener failed: \n {ex}")
-            if context:
-                context.set_code(grpc.StatusCode.INTERNAL)
-                context.set_details(f"{ex}")
-            return pb2.req_status()
+                if not traddr:
+                    raise Exception(f"unknown traddr")
+
+                return_string = self.spdk_rpc.nvmf.nvmf_subsystem_add_listener(
+                    self.spdk_rpc_client,
+                    nqn=request.nqn,
+                    trtype=request.trtype,
+                    traddr=traddr,
+                    trsvcid=request.trsvcid,
+                    adrfam=request.adrfam,
+                )
+                self.logger.info(f"Status of add listener: {return_string}")
+            except Exception as ex:
+                self.logger.error(f"Add Listener failed: \n {ex}")
+                if context:
+                    context.set_code(grpc.StatusCode.INTERNAL)
+                    context.set_details(f"{ex}")
+                return pb2.req_status()
 
         if context:
             # Update persistent configuration
```

Comment:
The persistent configuration is updated on gateway name mismatch with the expectation that all other gateways in the gateway group would see the change and attempt to execute the request, right?

If so, did you intend this update to be unconditional?  E.g. currently if a user specifies `gateway` but omits `traddr`, in case the handling gateway is the target gateway, the request is failed.  However, in case the handling gateway is not the target gateway, the persistent config is updated with an invalid record (which would be failed later by the target gateway, I assume).

Reply:
hope, it is better now